cmd/clusterctl/client/tree/discovery.go

```diff
@@ -268,20 +272,26 @@ func addMachineDeploymentToObjectTree(ctx context.Context, c client.Client, clus
 	return nil
 }
 
-func addMachinePoolsToObjectTree(ctx context.Context, c client.Client, namespace string, workers *unstructured.Unstructured, machinePoolList *expv1.MachinePoolList, tree *ObjectTree) {
+func addMachinePoolsToObjectTree(ctx context.Context, c client.Client, namespace string, workers *unstructured.Unstructured, machinePoolList *expv1.MachinePoolList, machinesList *clusterv1.MachineList, tree *ObjectTree, addMachineFunc func(parent client.Object, m *clusterv1.Machine)) {
```

Comment:
I just noticed that InfrastructureRef must be visible only if options.ShowTemplates is set (this will be consistent with what we are doing for MD) 
NOTE: in the case of MP this does not apply to Bootstrap.ConfigRef because it is not a template

Reply:
I'm handling the `mp.Spec.Template.Spec.Bootstrap.ConfigRef` and `mp.Spec.Template.Spec.InfrastructureRef` the same way we handle it for Machines since Machine Deployments don't have directly associated infra resources. Do you want to keep it like this then?

Reply:
Unfortunately, MP API is very confusing because references to infra and bootstrap ref are under a template struct, so technically they are templates to be used for Machines as we do in MD/MS, but in reality, those fields are used to reference objects used for the MP itself, and this makes them equals to the objects referenced by a Machine...

This will be confusing no matter what we do, so I think we can stick to your solution as well for now.

Feel free to resolve this, but I really hope we can find a better UX for this API before MP graduation